From: Padmanabh Ratnakar <padmanabh.ratnakar@broadcom.com>
Date: Thu, 18 Feb 2016 03:09:34 +0530
Subject: [PATCH 23/28] be2net: Fix pcie error recovery in case of NIC+RoCE
 adapters
Patch-mainline: v4.6-rc1
Git-commit: 68f227930b887ef9f18065f15a17aa9dc1ac2d31
References: bsc#963844 FATE#320192

Interrupts registered by RoCE driver are not unregistered when
msix interrupts are disabled during error recovery causing a
crash. Detach the adapter instance from RoCE driver when error
is detected to complete the cleanup. Attach the driver again after
the adapter is recovered from error.

Signed-off-by: Padmanabh Ratnakar <padmanabh.ratnakar@broadcom.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
Acked-by: David Chang <dchang@suse.com>
---
 drivers/net/ethernet/emulex/benet/be_main.c |    4 ++++
 1 file changed, 4 insertions(+)

--- a/drivers/net/ethernet/emulex/benet/be_main.c
+++ b/drivers/net/ethernet/emulex/benet/be_main.c
@@ -5456,6 +5456,8 @@ static pci_ers_result_t be_eeh_err_detec
 
 	dev_err(&adapter->pdev->dev, "EEH error detected\n");
 
+	be_roce_dev_remove(adapter);
+
 	if (!be_check_error(adapter, BE_ERROR_EEH)) {
 		be_set_error(adapter, BE_ERROR_EEH);
 
@@ -5520,6 +5522,8 @@ static void be_eeh_resume(struct pci_dev
 	if (status)
 		goto err;
 
+	be_roce_dev_add(adapter);
+
 	be_schedule_err_detection(adapter, ERR_DETECTION_DELAY);
 	return;
 err:
